cmd/pulumi-test-language/interface.go

```diff
@@ -0,0 +1,861 @@
+	Cancel()
+
+	// Done awaits the test servers termination, and returns any errors that result.
+	Done() error
+}
+
+func Start(ctx context.Context) (LanguageTestServer, error) {
+	// New up an engine RPC server.
+	server := &languageTestServer{
+		ctx:    ctx,
+		cancel: make(chan bool),
+	}
+
+	// Fire up a gRPC server and start listening for incomings.
+	port, done, err := rpcutil.Serve(0, server.cancel, []func(*grpc.Server) error{
+		func(srv *grpc.Server) error {
+			testingrpc.RegisterLanguageTestServer(srv, server)
+			return nil
+		},
+	}, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	server.addr = fmt.Sprintf("127.0.0.1:%d", port)
+	server.done = done
+
+	return server, nil
+}
+
+// languageTestServer is the server side of the language testing RPC machinery.
+type languageTestServer struct {
+	testingrpc.UnsafeLanguageTestServer
+
+	ctx    context.Context
+	cancel chan bool
+	done   chan error
+	addr   string
+}
+
+func (eng *languageTestServer) Address() string {
+	return eng.addr
+}
+
+func (eng *languageTestServer) Cancel() {
+	eng.cancel <- true
+}
+
+func (eng *languageTestServer) Done() error {
+	return <-eng.done
+}
+
+// A providerLoader is a schema loader that loads schemas from a given set of providers.
+type providerLoader struct {
+	providers []plugin.Provider
+}
+
+func (l *providerLoader) LoadPackageReference(pkg string, version *semver.Version) (schema.PackageReference, error) {
+	// Find the provider with the given package name
+	var provider plugin.Provider
+	for _, p := range l.providers {
+		if string(p.Pkg()) == pkg {
+			info, err := p.GetPluginInfo()
+			if err != nil {
+				return nil, fmt.Errorf("get plugin info for %s: %w", pkg, err)
+			}
+
+			if version == nil || (info.Version != nil && version.EQ(*info.Version)) {
+				provider = p
+				break
+			}
+		}
+	}
+
+	if provider == nil {
+		return nil, fmt.Errorf("could not load schema for %s, provider not known", pkg)
+	}
+
+	jsonSchema, err := provider.GetSchema(0)
+	if err != nil {
+		return nil, fmt.Errorf("get schema for %s: %w", pkg, err)
+	}
+
+	var spec schema.PartialPackageSpec
+	if _, err := json.Parse(jsonSchema, &spec, json.ZeroCopy); err != nil {
+		return nil, err
+	}
+
+	p, err := schema.ImportPartialSpec(spec, nil, l)
+	if err != nil {
+		return nil, err
+	}
+
+	return p, nil
+}
+
+func (l *providerLoader) LoadPackage(pkg string, version *semver.Version) (*schema.Package, error) {
+	ref, err := l.LoadPackageReference(pkg, version)
+	if err != nil {
+		return nil, err
+	}
+	return ref.Definition()
+}
+
+func (eng *languageTestServer) GetLanguageTests(
+	ctx context.Context,
+	req *testingrpc.GetLanguageTestsRequest,
+) (*testingrpc.GetLanguageTestsResponse, error) {
+	tests := make([]string, 0, len(languageTests))
+	for testName := range languageTests {
+		// Don't return internal tests
+		if strings.HasPrefix(testName, "internal-") {
+			continue
+		}
+		tests = append(tests, testName)
+	}
```

Comment:
languageTests is a map, so the order of items in `tests` will be non-deterministic. Should this have a `sort.Strings` at the end?

Reply:
Order shouldn't really matter, but if it did matter we'd want declaration order. Could either change this to a list (but then we get slower lookups) or add a counter to each item in the array and sort by that here.